{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import dask as dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20, 7)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the manifest\n",
    "mani = pd.read_csv('manifests/sample_manifest.csv')\n",
    "mani.shape\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>motif</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0fac0f32-9377-5739-901d-94baa2d17f0b</td>\n",
       "      <td>10</td>\n",
       "      <td>GCC</td>\n",
       "      <td>X:147582151-147582211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0fac0f32-9377-5739-901d-94baa2d17f0b</td>\n",
       "      <td>21</td>\n",
       "      <td>GCA</td>\n",
       "      <td>X:66765158-66765227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0fac0f32-9377-5739-901d-94baa2d17f0b</td>\n",
       "      <td>19/20</td>\n",
       "      <td>CAG</td>\n",
       "      <td>12:7045879-7045936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0fac0f32-9377-5739-901d-94baa2d17f0b</td>\n",
       "      <td>29/31</td>\n",
       "      <td>TGC</td>\n",
       "      <td>6:16327864-16327954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0fac0f32-9377-5739-901d-94baa2d17f0b</td>\n",
       "      <td>14/14</td>\n",
       "      <td>ATTCT</td>\n",
       "      <td>22:46191234-46191304</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 sample genotype  motif                 region\n",
       "0  0fac0f32-9377-5739-901d-94baa2d17f0b       10    GCC  X:147582151-147582211\n",
       "1  0fac0f32-9377-5739-901d-94baa2d17f0b       21    GCA    X:66765158-66765227\n",
       "2  0fac0f32-9377-5739-901d-94baa2d17f0b    19/20    CAG     12:7045879-7045936\n",
       "3  0fac0f32-9377-5739-901d-94baa2d17f0b    29/31    TGC    6:16327864-16327954\n",
       "4  0fac0f32-9377-5739-901d-94baa2d17f0b    14/14  ATTCT   22:46191234-46191304"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "examdf = pd.read_json('data/ndjson_examples/exmaples/0fac0f32-9377-5739-901d-94baa2d17f0b.ndjson', lines=True)\n",
    "examdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data and split the genotype column into two numeric columns\n",
    "def process_dataframe(path):\n",
    "    # Load the data\n",
    "    df = pd.read_json(path, lines=True)\n",
    "\n",
    "    # Apply add_slash to the 'genotype' column\n",
    "    df['genotype'] = df['genotype'].astype('string').apply(add_slash)\n",
    "\n",
    "    # Split the 'genotype' column into two\n",
    "    df[['genotype1', 'genotype2']] = df['genotype'].str.split('/', expand=True)\n",
    "\n",
    "    # Convert genotypes to numeric, errors='coerce' will convert invalid parsing to NaN\n",
    "    df[['genotype1', 'genotype2']] = df[['genotype1', 'genotype2']].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "# if there is no slash in the string, make it 'orginal/original'\n",
    "def add_slash(string):\n",
    "    if pd.isna(string):\n",
    "        return string\n",
    "    elif '/' not in string:\n",
    "        return '{}/{}'.format(string, string)\n",
    "    else:\n",
    "        return string\n",
    "    \n",
    "    \n",
    "# Load the data and subtract the genotype columns\n",
    "def load_and_diff(sample):\n",
    "    path_format = 'data/ndjson_examples/exmaples/{id}.ndjson'\n",
    "    # Use the helper function to process the dataframes\n",
    "    dfn = process_dataframe(path_format.format(id=sample['control_object_id']))\n",
    "    dft = process_dataframe(path_format.format(id=sample['case_object_id']))\n",
    "\n",
    "    # Subtract genotype1 and genotype2 columns in both dataframes\n",
    "    diff1 = dfn['genotype1'].subtract(dft['genotype1'], fill_value=0)\n",
    "    diff2 = dfn['genotype2'].subtract(dft['genotype2'], fill_value=0)\n",
    "\n",
    "    # Create a new DataFrame from these Series\n",
    "    diff_df = pd.DataFrame([diff1, diff2])\n",
    "\n",
    "    return diff_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "icgc_donor_id                                          DO46827\n",
      "case_object_id            d1dbd41b-c82d-5c9d-b404-07f97352e6f1\n",
      "control_object_id         b4f8d104-4de4-5ceb-aee0-08f599351f00\n",
      "case_file_name            72fa6dda655cffe29e8ae26357ec3069.bam\n",
      "control_file_name         6e7e2e09c1b59afdda8266a5d2706de2.bam\n",
      "sex                                                       male\n",
      "histology_abbreviation                              Kidney-RCC\n",
      "Name: 1, dtype: object\n"
     ]
    }
   ],
   "source": [
    "test = mani.iloc[1]\n",
    "print(test)\n",
    "dfn = pd.read_json('data/ndjson_examples/exmaples/{id}.ndjson'.format(id = test['control_object_id']), lines=True)\n",
    "dft = pd.read_json('data/ndjson_examples/exmaples/{id}.ndjson'.format(id = test['case_object_id']), lines=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample      0\n",
       "genotype    0\n",
       "motif       0\n",
       "region      0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfn.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>170856</th>\n",
       "      <th>170857</th>\n",
       "      <th>170858</th>\n",
       "      <th>170859</th>\n",
       "      <th>170860</th>\n",
       "      <th>170861</th>\n",
       "      <th>170862</th>\n",
       "      <th>170863</th>\n",
       "      <th>170864</th>\n",
       "      <th>170865</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>genotype1</th>\n",
       "      <td>-5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype2</th>\n",
       "      <td>-5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 170866 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0       1       2       3       4       5       6       7       \\\n",
       "genotype1    -5.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "genotype2    -5.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "           8       9       ...  170856  170857  170858  170859  170860  \\\n",
       "genotype1    -3.0     0.0  ...     0.0     0.0     0.0     1.0     0.0   \n",
       "genotype2    -3.0     0.0  ...     0.0     0.0     0.0     1.0     0.0   \n",
       "\n",
       "           170861  170862  170863  170864  170865  \n",
       "genotype1     0.0     0.0     0.0     0.0     0.0  \n",
       "genotype2     0.0     0.0     0.0     0.0     0.0  \n",
       "\n",
       "[2 rows x 170866 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>motif</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>11/11</td>\n",
       "      <td>GCC</td>\n",
       "      <td>X:147582151-147582211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>19/19</td>\n",
       "      <td>GCA</td>\n",
       "      <td>X:66765158-66765227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>20/22</td>\n",
       "      <td>CAG</td>\n",
       "      <td>12:7045879-7045936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>30/32</td>\n",
       "      <td>TGC</td>\n",
       "      <td>6:16327864-16327954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>13/14</td>\n",
       "      <td>ATTCT</td>\n",
       "      <td>22:46191234-46191304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170861</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>12/12</td>\n",
       "      <td>AC</td>\n",
       "      <td>X:99177272-99177300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170862</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>12/12</td>\n",
       "      <td>AC</td>\n",
       "      <td>X:99179509-99179533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170863</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>8/8</td>\n",
       "      <td>GATA</td>\n",
       "      <td>X:99183991-99184019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170864</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>16/16</td>\n",
       "      <td>TG</td>\n",
       "      <td>X:99224659-99224703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170865</th>\n",
       "      <td>b4f8d104-4de4-5ceb-aee0-08f599351f00</td>\n",
       "      <td>3/3</td>\n",
       "      <td>AATAA</td>\n",
       "      <td>X:99234955-99234975</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170866 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      sample genotype  motif  \\\n",
       "0       b4f8d104-4de4-5ceb-aee0-08f599351f00    11/11    GCC   \n",
       "1       b4f8d104-4de4-5ceb-aee0-08f599351f00    19/19    GCA   \n",
       "2       b4f8d104-4de4-5ceb-aee0-08f599351f00    20/22    CAG   \n",
       "3       b4f8d104-4de4-5ceb-aee0-08f599351f00    30/32    TGC   \n",
       "4       b4f8d104-4de4-5ceb-aee0-08f599351f00    13/14  ATTCT   \n",
       "...                                      ...      ...    ...   \n",
       "170861  b4f8d104-4de4-5ceb-aee0-08f599351f00    12/12     AC   \n",
       "170862  b4f8d104-4de4-5ceb-aee0-08f599351f00    12/12     AC   \n",
       "170863  b4f8d104-4de4-5ceb-aee0-08f599351f00      8/8   GATA   \n",
       "170864  b4f8d104-4de4-5ceb-aee0-08f599351f00    16/16     TG   \n",
       "170865  b4f8d104-4de4-5ceb-aee0-08f599351f00      3/3  AATAA   \n",
       "\n",
       "                       region  \n",
       "0       X:147582151-147582211  \n",
       "1         X:66765158-66765227  \n",
       "2          12:7045879-7045936  \n",
       "3         6:16327864-16327954  \n",
       "4        22:46191234-46191304  \n",
       "...                       ...  \n",
       "170861    X:99177272-99177300  \n",
       "170862    X:99179509-99179533  \n",
       "170863    X:99183991-99184019  \n",
       "170864    X:99224659-99224703  \n",
       "170865    X:99234955-99234975  \n",
       "\n",
       "[170866 rows x 4 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "difftest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
